pkg/operator/controllers/node/node_controller.go

```diff
@@ -0,0 +1,155 @@
+package node
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/kubectl/pkg/drain"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/ready"
+)
+
+const (
+	annotationCurrentConfig  = "machineconfiguration.openshift.io/currentConfig"
+	annotationDesiredConfig  = "machineconfiguration.openshift.io/desiredConfig"
+	annotationState          = "machineconfiguration.openshift.io/state"
+	annotationDrainStartTime = "aro.openshift.io/drainStartTime"
+	stateWorking             = "Working"
+	gracePeriod              = 10 * time.Minute
+)
+
+// NodeReconciler spots nodes that look like they're stuck upgrading.  When this
+// happens, it tries to drain them disabling eviction (so PDBs don't count).
+type NodeReconciler struct {
+	log           *logrus.Entry
+	kubernetescli kubernetes.Interface
+}
+
+func NewNodeReconciler(log *logrus.Entry, kubernetescli kubernetes.Interface) *NodeReconciler {
+	return &NodeReconciler{
+		log:           log,
+		kubernetescli: kubernetescli,
+	}
+}
+
+func (r *NodeReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
+	// TODO(mj): controller-runtime master fixes the need for this (https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L93) but it's not yet released.
+	ctx := context.Background()
+
+	node, err := r.kubernetescli.CoreV1().Nodes().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		r.log.Error(err)
+		return reconcile.Result{}, err
+	}
+
+	if !isDraining(node) {
+		// we're not draining: ensure our annotation is not set and return
+		if getAnnotation(&node.ObjectMeta, annotationDrainStartTime) == "" {
+			return reconcile.Result{}, nil
+		}
+
+		delete(node.Annotations, annotationDrainStartTime)
+
+		_, err = r.kubernetescli.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+		if err != nil {
+			r.log.Error(err)
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// we're draining: ensure our annotation is set
+	t, err := time.Parse(time.RFC3339, getAnnotation(&node.ObjectMeta, annotationDrainStartTime))
+	if err != nil {
+		t = time.Now().UTC()
+		setAnnotation(&node.ObjectMeta, annotationDrainStartTime, t.Format(time.RFC3339))
+
+		node, err = r.kubernetescli.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+		if err != nil {
+			r.log.Error(err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// if our deadline hasn't expired, requeue ourselves and return
+	deadline := t.Add(gracePeriod)
+	now := time.Now()
```

Comment:
Shouldn't this be UTC time?  Just to be consistent with above?

Reply:
Will double check, but this shouldn't matter.

Reply:
It doesn't.